### ISSUE TITLE
Large number of security warnings

### ISSUE THREAD

**Issue by Itach1Uchixa** [LonamiWebs/Telethon#1476]
**Checklist**
* [x] The error is in the library's code, and not in my own.
* [x] I have searched for this issue before posting it and there isn't a duplicate.
* [x] I ran `pip install -U https://github.com/LonamiWebs/Telethon/archive/master.zip` and triggered the bug in the latest version.

Our software works with ~100 Telegram sessions at the same time. Doing so causes API calls to output, below warning when 100 sessions do something at once asynchronously. Output numbers vary between 50 to 200 or more sometimes.

**Warning**
```
WARNING:telethon.network.mtprotosender:Security error while unpacking a received message: Server replied with a wrong session ID
```

**Comment by Lonami** [comment]
Unfortunately, I don't know why these occur, how to reproduce it, or whether they actually matter or can be safely ignored and the client just restarted.

**Comment by Itach1Uchixa** [comment]
We did runtime tests and noticed that when we reduce the number of connected sessions, this warning pops out in the rare occasions

**Comment by Lonami** [comment]
If I flip a random bit here:
https://github.com/LonamiWebs/Telethon/blob/fc07e6bba78da2f3ed0fd9e94290aef4609dbd80/telethon/network/mtprotosender.py#L460

Then I get
```
Security error while unpacking a received message: Received msg_key doesn't match with expected one
```

However your error is different and can only occur if the `msg_key` is correct:

https://github.com/LonamiWebs/Telethon/blob/fc07e6bba78da2f3ed0fd9e94290aef4609dbd80/telethon/network/mtprotostate.py#L126-L139

(Furthermore, the default connection mode has an additional CRC checksum so it would be *very* rare for it to have bits flipped often enough, but anyway your error is different so we discard this possibility).

I wonder if the payload is actually valid, but we never get to see this because we raise early. It could be interesting to patch that code to not raise (or raise later) to see what Telegram is sending that doesn't belong to our session.

Anyway, I still don't know why this happens or what the correct way to act is.

**Comment by ZIVVAL** [comment]
same here - having large number of security warnings 
telethon.network.mtprotosender: Security error while unpacking a received message: Server replied with a wrong session ID

**Comment by wind-hx** [comment]
I have the same problem

**Comment by wind-hx** [comment]
SecurityError: Received msg_key doesn't match with expected one

**Comment by migcm** [comment]
I have the same problem.

**Comment by mibj** [comment]
Hi 

just to report I have the same problem :
`WARNING:telethon.network.mtprotosender:Security`` error while unpacking a received message: Server replied with a wrong session ID`

270 session
Changed it to 200 it happens less

**Comment by Lonami** [comment]
Perhaps Telegram "gets confused" with so many connections. Do things still work if the logs are ignored, or does it not function after the warnings are logged?

**Comment by mibj** [comment]
the code still works , It doesn't crash , but I don't get the whole file but a part of it . Many files aren't complete

**Comment by Lonami** [comment]
You download files and, when this happens, the files are incomplete?

**Comment by mibj** [comment]
Yes some files but not all. See the code below.
maybe I need a queues or reduce the number of connections
`
import asyncio 
import sys
from telethon import TelegramClient,events
from telethon.tl import functions, types
from datetime import datetime
from telethon.tl.functions.messages import ImportChatInviteRequest

file_path = ''
channel1  = "" 
channel2  = "" 
api_id    = ""  
api_hash  = ""
#>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
DEBUG = True 
if DEBUG:
    import logging
    logging.basicConfig(format='[%(levelname) 5s/%(asctime)s] %(name)s: %(message)s',
        level=logging.WARNING)

client = TelegramClient('Test', api_id, api_hash)
client.start()

def callback(current, total):
     print('Downloaded', current, 'out of', total,
           'bytes: {:.2%}'.format(current / total))

@client.on(events.NewMessage(chats=(channel1,channel2))) 
async def my_event_handler(event):
        await event.download_media(file_path,progress_callback=callback)

client.run_until_disconnected()
`

**Comment by moritz-fenske** [comment]
Sam error while i have around 5 connections.

**Comment by BroLabsOfficial** [comment]
I have the same problem

**Comment by Sladerix** [comment]
Same error, but i receive the message anyway

**Comment by buddhhu** [comment]
String session gets expired after some warnings also

**Comment by Lonami** [comment]
Do things work if you replace this line:

https://github.com/LonamiWebs/Telethon/blob/d44928c27b4a34e396414a79966726c8877c023b/telethon/network/mtprotostate.py#L159

with `pass`?

**Comment by New-dev0** [comment]
> Do things work if you replace this line:
> 
> https://github.com/LonamiWebs/Telethon/blob/d44928c27b4a34e396414a79966726c8877c023b/telethon/network/mtprotostate.py#L159
> 
> with `pass`?

You are telling to Try with Expecting and passing SecurityError ? Right no?

**Comment by buddhhu** [comment]
> Do things work if you replace this line:
> 
> https://github.com/LonamiWebs/Telethon/blob/d44928c27b4a34e396414a79966726c8877c023b/telethon/network/mtprotostate.py#L159
> 
> with `pass`?

It won't. If u pass it then it will directly create AuthDuplicatedError without that wrong session id warning.

**Comment by Lonami** [comment]
`AuthKeyDuplicatedError` occurs when you use the exact same session under two different IP addresses. You cannot do this. You need to create a different session.

**Comment by buddhhu** [comment]
> `AuthKeyDuplicatedError` occurs when you use the exact same session under two different IP addresses. You cannot do this. You need to create a different session.

Yes but after that wrong session id waring it expires after some event is triggered

**Comment by Lonami** [comment]
It probably expires because the session is being used from two places. Is this the case?

**Comment by buddhhu** [comment]
> It probably expires because the session is being used from two places. Is this the case?

Maybe not. Many users are using, some doesn't face this (me present), some doesn't even face the wrong session id (I was getting warning), and for some it just expires the session after some warning (many in present).

**Comment by Sladerix** [comment]
Same problem here, but it works anyway

**Comment by Thoufak** [comment]
> I wonder if the payload is actually valid, but we never get to see this because we raise early. It could be interesting to patch that code to not raise (or raise later) to see what Telegram is sending that doesn't belong to our session.

@Lonami, I have done this. I myself encounter this issue quite often when working with multiple accounts simultaneously. I modified the Telethon's code to raise the exception later and see the payload. I organised the data into a json file for easier analysis. It's a list of 716 entries which have the following fields:
- expected_id – see the screenshot for explanation.
- actual_id – see the screenshot for explanation.
- remote_msg_id – original variable.
- remote_sequence – original variable.
- obj - original variable on which I called the json() method.

![Screenshot 2021-06-15 at 11 07 58](https://user-images.githubusercontent.com/40132582/122018809-e17c3180-cdcb-11eb-9346-0b8084dd2b50.png)

An approximate description of my case: 
100+ accounts being connected for a few hours. After some point the console explodes with a lot of these warning. At first, the script operates normally, but then it hangs (but the warnings keep coming).

Please find the json file with all the data attached.
Is this enough for you to analyse the problem? I can provide more data if you tell me what's needed.

[results.txt](https://github.com/LonamiWebs/Telethon/files/6654007/results.txt)

**Comment by DenisTsukanov** [comment]
Same problem, I'v started facing it, as soon as I started adding more accounts to work, all of them connect via proxy, also a bunch of these errors occur after some time of work, depends on amount of accounts. Usually it can be suspended up to 45 minutes, but than... it's a disaster

**Comment by Lonami** [comment]
> Is this enough for you to analyse the problem

Thanks for the effort @Thoufak. It's interesting how the `actual_id` are repeated (and not even immediately one after the other). Beyond the overwhelming majority of `RpcResult`, `NewSession` and `MsgAck` also seem to dominate. There's also "expected IDs" in the "actual IDs". So it really seems Telegram is "getting confused" with so many sessions and, because it's the same client after all, sends everything to the same address, maybe?

I don't know if this can be easily solved. One option would be to have a map of session ID to the clients so they can be dispatched correctly, but this would be a radical change in how the code is structured and how things work (unlikely to happen). Maybe it's `asyncio`-related, reusing connections. Maybe it's something Telethon is doing wrong with regards to sessions.

Are all the clients being ran in the same script, and is the problem solved if there was a separate process per session?

**Comment by simplPart** [comment]
> Are all the clients being ran in the same script, and is the problem solved if there was a separate process per session?

i remember i had the same problem.  I decided to simply divide the clients into several threads and have its own loop for each thread - it helped.  but now one loop for many threads has been added and errors have appeared again... in the loop I use proactoreventloop - only this selector allows you to use a proxy

**Comment by Lonami** [comment]
In that case, I'm really leaning towards the hypothesis of `asyncio.open_connection` maintaining some sort of network pool which gets reused, though I haven't confirmed it or investigated.

I do wonder if the old Telethon's approach of using `socket` directly suffered from this problem (probably too old by now to bother testing those).

**Comment by BelgenOp** [comment]
@Lonami , do you got solution for session warnings?

**Comment by ehtec** [comment]
Any updates regarding this? Or is there a workaround? It would be sufficient for me if the request is retried on the warning. Or is this already done? Otherwise I will have lots of trouble with incomplete files or missing messages

**Comment by ehtec** [comment]
I have had a look at this file:

https://github.com/LonamiWebs/Telethon/blob/d44928c27b4a34e396414a79966726c8877c023b/telethon/network/mtprotosender.py

It says the message is not safe and should be ignored. I am not receiving any messages real-time, I use `iter_messages`. So how can I catch the Security Error properly and retry fetching the messages?

I noticed that this error mainly occurs when I run another script in parallel which does not use Telethon at all, but puts a high load on the hard drive and the PostgreSQL database the main script is also interacting with. Any idea why this is happening? I already tried setting the disk priority of the other process as low as possible.

**Comment by Lonami** [comment]
@ehtec my comments above (https://github.com/LonamiWebs/Telethon/issues/1476#issuecomment-862102252 and https://github.com/LonamiWebs/Telethon/issues/1476#issuecomment-939519939) are the most probable cause and path of action, however, I've been busy with other things.

**Comment by buddhhu** [comment]
`client.disconnect()` and `client.connect()` still doesn't solve this issue but restarting the script solves sometimes 🤔

**Comment by ehtec** [comment]
@Lonami Thanks for the reply, have not seen it somehow. What is the current handling of the error, is the message discarded or the fetch retried? I can live with retry until success, but to lose messages would be really bad.

**Comment by graysuit** [comment]
I rarely this warnings:
`Security error while unpacking a received message: Server replied with a wrong session ID`
I do long time flirt with telegram api. and in b/w it raises sometimes. But code works fine.

my imports are:
```python
from telethon import TelegramClient, events, sync, connection, functions, types
from telethon.tl.types import InputPhoneContact
from telethon.errors import FloodWaitError
```

It raises with both MTPROTO proxy and without.
Warning happens mostly when dealing with `InputPhoneContact`, `ImportContactsRequest`,`FloodWaitError`.

**Comment by Krishna26-master** [comment]
Change api variables to remove warnings

**Comment by SharkAF2** [comment]
> Change api variables to remove warnings

How would you do that?

**Comment by SharkAF2** [comment]
> @ehtec my comments above ([#1476 (comment)](https://github.com/LonamiWebs/Telethon/issues/1476#issuecomment-862102252) and [#1476 (comment)](https://github.com/LonamiWebs/Telethon/issues/1476#issuecomment-939519939)) are the most probable cause and path of action, however, I've been busy with other things.

Any updates?

**Comment by keyiflerolsun** [comment]
![image](https://user-images.githubusercontent.com/57468649/153985979-1802f80e-5bcf-4406-99ae-40ba83408f21.png)

this not solved :)

**Comment by wind-hx** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by lgg** [comment]
I faced this issue after I changed group to supergroup (history: visible).

Deleting `bot.session` file and restarting bot - helped me to fix this issue.

**Comment by w1ld32** [comment]
I get this problem when I try to register a new account using a voice call. Before sending a voice message, you need to wait 180 seconds, after waiting and trying to resend the code, I get this error and do not receive a call.

**Comment by wind-hx** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
